File: stack-queues/739.py
# Tc = O(n)
# Sc = O(n)
class Solution:
    def dailyTemperatures(self, temperatures):
        n = len(temperatures)
        st = []
        result = [0] * n
        
        for i in range(n - 1, -1, -1):
            while st and temperatures[i] >= temperatures[st[-1]]:
                st.pop()
            if not st:
                result[i] = 0
            else:
                result[i] = st[-1] - i
            st.append(i)
        
        return result

File: stack-queues/739.js
class Solution {
    dailyTemperatures(temp){
        const n = temp.length
        const stack = []
        const result = new Array(n).fill(0)

        for(let i = n-1; i>=0; i--){
            while(stack.length && temp[i] >= temp[stack[stack.length-1]]){
                stack.pop()
            }
            if(stack.length === 0){
                result[i] = 0
            }else{
                result[i] = stack[stack.length - 1] - i
            }
            stack.push(i)
        }
        return result
    }
}